File: scripts/shot.mjs
/* Screenshot the app from Node, using the Playwright that is already in
   node_modules. Companion to scripts/shot.py, which needs the Python package —
   and that one refuses to install on the Python 3.9 here (greenlet has no
   wheel). Same idea, fewer states: boot the test server, drive the app past the
   language screen, save PNGs, and report anything the console complained about.

     node scripts/shot.mjs                       list + detail, tablet
     node scripts/shot.mjs --vp phone --out dir

   This is a *look at it* tool, not a test — it asserts nothing. Read the PNGs. */
import { chromium } from "playwright-core";
import { mkdirSync } from "node:fs";
import { resolve } from "node:path";

const arg = (flag, dflt) => {
  const i = process.argv.indexOf(flag);
  return i > -1 && process.argv[i + 1] ? process.argv[i + 1] : dflt;
};

const VIEWPORTS = {
  tablet: { width: 1024, height: 768, isMobile: false },
  phone: { width: 390, height: 844, deviceScaleFactor: 2, isMobile: true },
  laptop: { width: 1440, height: 900, isMobile: false }
};

const vpName = arg("--vp", "tablet");
const out = resolve(arg("--out", "shots"));
const base = process.env.BASE || "http://127.0.0.1:4173";

mkdirSync(out, { recursive: true });

const browser = await chromium.launch();
const ctx = await browser.newContext({
  ...VIEWPORTS[vpName],
  hasTouch: vpName !== "laptop",
  reducedMotion: "reduce"
});
const page = await ctx.newPage();

const problems = [];
page.on("pageerror", (e) => problems.push("page error: " + e));
page.on("console", (m) => { if (m.type() === "error") problems.push("console: " + m.text()); });
page.on("response", (r) => { if (r.status() >= 400) problems.push(`${r.status()} ${r.url()}`); });

await page.goto(base + "/index.html", { waitUntil: "load" });
await page.screenshot({ path: resolve(out, `${vpName}-start.png`) });

await page.evaluate(() => localStorage.setItem("theatrium-lang", "hr"));
await page.goto(base + "/index.html", { waitUntil: "load" });
await page.waitForFunction(() => typeof DATA !== "undefined" && !!DATA);
const enter = page.locator("#story-enter");
if (await enter.isVisible().catch(() => false)) await enter.click();
await page.waitForTimeout(400);
await page.screenshot({ path: resolve(out, `${vpName}-list.png`) });

const counts = await page.evaluate(() => ({
  sections: DATA.sections.length,
  listings: DATA.sections.flatMap((s) => s.categories.flatMap((c) =>
    c.groups.flatMap((g) => g.items))).length,
  first: DATA.sections[0].categories[0].groups[0].items[0].name
}));

await page.locator("#content .item").first().click();
await page.waitForTimeout(400);
await page.screenshot({ path: resolve(out, `${vpName}-detail.png`) });
const sheet = await page.locator(".detail-name").innerText().catch(() => "(none)");

console.log(`${vpName}: ${counts.listings} listings in ${counts.sections} sections`);
console.log(`  first wine: ${counts.first}   detail sheet opened: ${sheet}`);
console.log(problems.length ? "  PROBLEMS:\n   " + problems.join("\n   ") : "  no console/page/network errors");
console.log("  " + out);

await browser.close();

File: scripts/shot.py
# -*- coding: utf-8 -*-
"""Screenshot the app in a real browser, so visual work stops being guesswork.

jsdom tells us the structure rendered; it never paints, so opacity, contrast,
crop and overlap are invisible to it. This boots a throwaway static server and a
headless Chromium, walks the app into each state worth looking at, and writes
PNGs to a directory.

    python scripts/shot.py                       # every state, every viewport
    python scripts/shot.py story                 # one state
    python scripts/shot.py story --vp tablet     # one state, one viewport
    python scripts/shot.py --out shots/before    # somewhere to diff against

States: start, story, list, detail, ikone, search.
Viewports: tablet 1024x768, laptop 1440x900, phone 390x844.

Needs `pip install playwright` and `playwright install chromium`.
"""
import argparse
import functools
import http.server
import os
import socketserver
import sys
import threading

from playwright.sync_api import sync_playwright

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
VIEWPORTS = {
    "tablet": dict(width=1024, height=768, is_mobile=False),
    "laptop": dict(width=1440, height=900, is_mobile=False),
    "phone": dict(width=390, height=844, is_mobile=True),
}


def serve():
    """A quiet static server on a free port, daemonised."""
    class Quiet(http.server.SimpleHTTPRequestHandler):
        def log_message(self, *a, **k):
            pass

    handler = functools.partial(Quiet, directory=ROOT)
    srv = socketserver.TCPServer(("127.0.0.1", 0), handler)
    threading.Thread(target=srv.serve_forever, daemon=True).start()
    return "http://127.0.0.1:%d" % srv.server_address[1]


def enter(page, base, lang="hr"):
    """Language screen → story splash. Returns with the splash on screen."""
    page.goto(base + "/index.html", wait_until="networkidle")
    page.evaluate("localStorage.clear()")
    page.goto(base + "/index.html", wait_until="networkidle")
    page.click('#lang-buttons button[data-lang="%s"]' % lang)
    page.wait_for_selector("#story-screen:not(.hidden)")
    page.wait_for_timeout(400)


def to_list(page):
    page.click("#story-enter")
    page.wait_for_selector("#app:not(.hidden)")
    page.wait_for_timeout(400)


STATES = {}


def state(fn):
    STATES[fn.__name__] = fn
    return fn


@state
def start(page, base):
    page.goto(base + "/index.html", wait_until="networkidle")
    page.evaluate("localStorage.clear()")
    page.goto(base + "/index.html", wait_until="networkidle")
    page.wait_for_selector("#start:not(.hidden)")
    page.wait_for_timeout(600)          # let the logo breathe animation settle


@state
def story(page, base):
    enter(page, base)


@state
def list(page, base):
    enter(page, base)
    to_list(page)


@state
def detail(page, base):
    enter(page, base)
    to_list(page)
    page.click(".item")
    page.wait_for_selector("#modal:not(.hidden)")
    page.wait_for_timeout(500)


@state
def ikone(page, base):
    enter(page, base)
    to_list(page)
    page.click("#pride-toggle")
    page.wait_for_timeout(500)


@state
def spirit(page, base):
    """A spirit card — a different set of fields and a different glass from a
    wine, so it needs looking at on its own."""
    enter(page, base)
    to_list(page)
    page.click("#search-toggle")
    page.fill("#search", "hampden")
    page.wait_for_timeout(400)
    page.click(".item.clickable")
    page.wait_for_selector("#modal:not(.hidden)")
    page.wait_for_timeout(500)


@state
def search(page, base):
    enter(page, base)
    to_list(page)
    page.click("#search-toggle")
    page.fill("#search", "barolo")
    page.wait_for_timeout(500)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("states", nargs="*", default=None,
                    help="states to shoot (default: all of %s)" % ", ".join(STATES))
    ap.add_argument("--vp", action="append", choices=sorted(VIEWPORTS),
                    help="viewport(s) to use (default: all)")
    ap.add_argument("--out", default="shots", help="output directory")
    ap.add_argument("--full", action="store_true", help="full-page instead of viewport-only")
    args = ap.parse_args()

    want = args.states or sorted(STATES)
    bad = [s for s in want if s not in STATES]
    if bad:
        sys.exit("unknown state(s): %s — pick from %s" % (", ".join(bad), ", ".join(sorted(STATES))))
    vps = args.vp or sorted(VIEWPORTS)

    os.makedirs(args.out, exist_ok=True)
    base = serve()
    written = []
    with sync_playwright() as p:
        # PW_CHROMIUM lets a machine whose Chromium build doesn't match the
        # installed playwright package point at the browser it does have,
        # instead of downloading a second ~120 MB copy.
        exe = os.environ.get("PW_CHROMIUM")
        browser = p.chromium.launch(executable_path=exe) if exe else p.chromium.launch()
        for vp in vps:
            cfg = dict(VIEWPORTS[vp])
            mobile = cfg.pop("is_mobile")
            ctx = browser.new_context(viewport=cfg, is_mobile=mobile,
                                      has_touch=mobile, device_scale_factor=2,
                                      reduced_motion="reduce")
            page = ctx.new_page()
            errs = []
            page.on("pageerror", lambda e: errs.append(str(e)))
            page.on("console", lambda m: errs.append("console.%s: %s" % (m.type, m.text))
                    if m.type == "error" else None)
            for name in want:
                STATES[name](page, base)
                path = os.path.join(args.out, "%s-%s.png" % (name, vp))
                page.screenshot(path=path, full_page=args.full)
                written.append((path, os.path.getsize(path)))
            ctx.close()
            for e in dict.fromkeys(errs):          # de-duplicated, order kept
                print("  ! %s [%s]" % (e, vp))
        browser.close()

    for path, size in written:
        print("%s  %.0f kB" % (path, size / 1024))


if __name__ == "__main__":
    main()
